v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+// Upload reads events from the buffer and uploads them to the configured client.
+// All the events currently in the buffer are read and written to a gzip compressed JSON array to create a chunk of data.
+// Each chunk is limited by the uploadSizeLimitBytes.
+func (b *eventBuffer) Upload(ctx context.Context) error {
+	b.upload.Lock()
+	defer b.upload.Unlock()
+
+	eventLen := len(b.buffer)
+	if eventLen == 0 {
+		return &bufferEmpty{}
+	}
+
+	encoder := newChunkEncoder(b.uploadSizeLimitBytes)
+
+	for range eventLen {
+		event := b.readEvent()
+		if event == nil {
+			break
+		}
+
+		var result [][]byte
+		if event.chunk != nil {
+			result = [][]byte{event.chunk}
+		} else {
+			serialized, err := b.processEvent(event.EventV1)
+			if err != nil {
+				b.logError("%v", err)
+				continue
+			}
+
+			result, err = encoder.WriteBytes(serialized)
+			if err != nil {
+				b.incrMetric(logEncodingFailureCounterName)
+				b.logError("encoding failure: %v, dropping event with decision ID: %v", err, event.DecisionID)
+				continue
+			}
+		}
+
+		if err := b.uploadChunks(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	// flush any chunks that didn't hit the upload limit
+	result, err := encoder.Flush()
+	if err != nil {
+		b.incrMetric(logEncodingFailureCounterName)
+		b.logError("encoding failure: %v", err)
+		return nil
+	}
+
+	if err := b.uploadChunks(ctx, result); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// uploadChunks attempts to upload multiple chunks to the configured client.
+// In case of failure all the events are added back to the buffer.
+func (b *eventBuffer) uploadChunks(ctx context.Context, result [][]byte) error {
+	var err error
+	for _, chunk := range result {
+		// only attempt an upload if there hasn't been a failure
+		if err == nil {
+			err = uploadChunk(ctx, b.client, b.uploadPath, chunk)
+		}
+
+		// if an upload failed, requeue the chunk
+		if err != nil {
+			b.push(&bufferItem{chunk: chunk})
+		}
+	}
+	return err
+}
+
+// readEvent does a nonblocking read from the event buffer
+func (b *eventBuffer) readEvent() *bufferItem {
+	select {
+	case event := <-b.buffer:
+		return event
+	default:
+		return nil
+	}
+}
+
+// processEvent serializes the event and determines if the ND cache needs to be dropped
+func (b *eventBuffer) processEvent(event *EventV1) ([]byte, error) {
+	serialized, err := json.Marshal(event)
+	if err != nil {
+		b.incrMetric(logEncodingFailureCounterName)
+		return nil, fmt.Errorf("encoding failure: %w, dropping event with decision ID: %v", err, event.DecisionID)
+	}
+
+	if int64(len(serialized)) >= b.uploadSizeLimitBytes {
+		if event.NDBuiltinCache == nil {
+			return nil, fmt.Errorf("upload event size (%d) exceeds upload_size_limit_bytes (%d), dropping event with decision ID: %v",
+				int64(len(serialized)), b.uploadSizeLimitBytes, event.DecisionID)
+		}
+
+		// Attempt to drop the ND cache to reduce size. If it is still too large, drop the event.
+		event.NDBuiltinCache = nil
```

Comment:
The [old buffer](https://github.com/open-policy-agent/opa/blob/5932f76bb25a1c0da107530ae014e650f464731d/v1/plugins/logs/plugin.go#L1067) drops the ND-cache based on failure to marshal to json, or on failure to add to chunk encoder. I wonder if that is what we should to here too. Under most circumstances,  `b.uploadSizeLimitBytes` is likely to be large enough to account for _several_ events.

Reply:
Good idea, changed so ND cache is dropped in case of marshal error as well.